File: cudf-sys/cpp/column.cpp
#include "column.hpp"

#include <cudf/column/column.hpp>
#include <cudf/column/column_view.hpp>

namespace cudf_rs {

Column::Column(std::unique_ptr<cudf::column> column)
    : column_(std::move(column)) {}

Column::~Column() = default;

cudf::size_type Column::size() const noexcept { return column_->size(); };

cudf::column_view const &Column::view() const {
  std::call_once(column_view_once_, [this] {
    column_view_ = std::make_unique<cudf::column_view>(column_->view());
  });
  return *column_view_;
}

} // namespace cudf_rs

File: cudf-sys/src/column.rs
#[cxx::bridge]
mod ffi {
    #[namespace = "cudf"]
    unsafe extern "C++" {
        include!("cudf/column/column.hpp");

        type column;
        fn has_nulls(self: &column) -> bool;

        type column_view;
        fn size(self: &column_view) -> i32;
    }

    #[namespace = "cudf_rs"]
    unsafe extern "C++" {
        include!("cudf-sys/include/column.hpp");

        type Column;

        fn size(self: &Column) -> i32;
        fn view(self: &Column) -> Result<&column_view>;

        // To support UniquePtr<Column>
        fn _column(_: UniquePtr<Column>);
    }
}

pub use ffi::*;

unsafe impl Send for ffi::Column {}
unsafe impl Sync for ffi::Column {}
